### PR TITLE
CNI: Modified CNI to be pass through for Azure baremetal scenario

### PR DESCRIPTION
**Pull request by msvik**
**Reason for Change**:
Adding support in CNI for "baremetal scenario" where when CNI is executed on baremetal nodes, it is pass through and invokes Azure's node service to configure container networking. 

**Issue Fixed**:
Added 

**Requirements**:

- [ ] includes documentation
- [ x] adds unit tests

### CODE REVIEW

File: cni/network/network.go [hunk 1]
```diff
@@ -791,6 +814,31 @@ func (plugin *netPlugin) Delete(args *cniSkel.CmdArgs) error {
...
+		telemetry.SendCNIMetric(&cniMetric, plugin.tb)
+	}()
+
+	log.Printf("Execution mode :%s", nwCfg.ExecutionMode)
+	if nwCfg.ExecutionMode == string(Baremetal) {
+
+		log.Printf("Baremetal mode. Calling vnet agent for delete container")
+
+		// set the flag to true for attempt. This will be used by deferred telemetry method to emit metrics
+		deleteTried = true
```
Comment (ramiro-gamarra) [inline cni/network/network.go:837]: Flags in defers are hard to follow (you need to scan the entire method to know where it could have changed). Shows better intent ("only send metrics after this code path") if we extract the deferred anonymous function out, and defer in places where needed, like

```go
func sendMetric(...) {...}

func delete() { 
    if nwCfg.ExecutionMode == string(Baremetal) {
        defer sendMetric(...)
        return plugin.nnsClient.DeleteContainerNetworking()
    }

    defer sendMetric(...)
    return plugin.nm.DeleteEndpoint
}
```

File: cni/network/plugin/main.go [hunk 2]
```diff
@@ -173,7 +174,7 @@ func main() {
 
 	cniReport.GetReport(pluginName, version, ipamQueryURL)
 
-	netPlugin, err := network.NewPlugin(name, &config)
+	netPlugin, err := network.NewPlugin(name, &config, &nnsgrpcclient.NnsGrpcClient{})
```
Comment (ramiro-gamarra) [inline cni/network/plugin/main.go:177]: Just a thought on package naming, it's very common in Go to have packages reflect domain, such that types within can have succinct names, without losing context from a callers perspective. You could have an `nns` package, and then types such as `nns.Client`

File: cni/nnsgrpcclient/nnsgrpcclient_test.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+	setup()
+	code := m.Run()
+	teardown()
+	os.Exit(code)
+}
+
+func setup() {
+	mockserver  = nnsmockserver.NewNnsMockServer()
+	go  mockserver.StartGrpcServer(nnsPort)
+	fmt.Printf("mock nns server started\n")
```
Comment (ramiro-gamarra) [inline cni/nnsgrpcclient/nnsgrpcclient_test.go:25]: `fmt.Println` if you don't need formatting directives, also saves you from having to write the newline. Same for below.

File: cni/nnsgrpcclient/nnsgrpcclient_test.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+// CNI ADD to add container to network
+func TestAddContainerNetworking(t *testing.T) {
+
+	client := &NnsGrpcClient{}
+	if err := client.AddContainerNetworking(
+		 context.Background(),
+		"sf_8e9961f4-5b4f-4b3c-a9ae-c3294b0d9681",
+		"testnwspace"); err != nil {
+       t.Fatalf("TestAddContainerNetworking failed: %v", err)
```
Comment (ramiro-gamarra) [inline cni/nnsgrpcclient/nnsgrpcclient_test.go:41]: formatting seems off. if you are using goland, you can enable file watchers and goimports (like gofmt but with added functionality)

File: cni/nnsgrpcclient/nnsgrpcclient_test.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+	}
+
+	fmt.Printf("TestAddContainerNetworking success\n")
```
Comment (ramiro-gamarra) [inline cni/nnsgrpcclient/nnsgrpcclient_test.go:44]: - you can use `t.Log` or `t.Logf` for logging within the test
- running the tests in verbose mode will tell you which pass and fail, so this seems a bit redundant. same for below

File: cni/nnsgrpcclient/nnsgrpcclient_test.go [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+func teardown() {
+	mockserver.StopGrpcServer()
+	fmt.Printf("mock nns server stopped\n")
+}
+
+// CNI ADD to add container to network
+func TestAddContainerNetworking(t *testing.T) {
+
+	client := &NnsGrpcClient{}
```
Comment (ramiro-gamarra) [inline cni/nnsgrpcclient/nnsgrpcclient_test.go:36]: how does the client know to hit the mock server?

File: cni/nnsgrpcclient/nnsgrpcclient_test.go [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+func TestAddContainerNetworkingFailure(t *testing.T) {
+
+	client := &NnsGrpcClient{}
+
+	var err error
+	if err = client.AddContainerNetworking(context.Background(), "testpod", "testnwspace"); err == nil {
+		t.Fatalf("TestAddContainerNetworkingFailure failed. Expected error but none returned\n")
+	}
+
+	if !strings.Contains(err.Error(), "Setup") {
```
Comment (ramiro-gamarra) [inline cni/nnsgrpcclient/nnsgrpcclient_test.go:71]: Checking the string representation of errors is brittle. It's better to have error types. In this case, does it matter what specific error was returned? Same for below

File: cni/nnsgrpcclient/nnsgrpcclient_test.go [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+func TestAddContainerNetworkingGrpcServerDown(t *testing.T) {
+
+	// shutdown server to simulate connection error
+	teardown()
+
+	// bring up the server again after 25 seconds
+	go func() {
+		time.Sleep(25 * time.Second)
+		setup()
+	}()
```
Comment (ramiro-gamarra) [inline cni/nnsgrpcclient/nnsgrpcclient_test.go:105]: how does this test work? does the client retry indefinitely? if it doesn't, then since the mock server is shared in many tests, this can leave set up broken for future tests. If it does, 25 seconds seems very long for this test to run (and for the client to retry). let me know if im missing something

File: cni/nnsgrpcclient/nnsgrpclient.go [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+	"context"
+	"fmt"
+	"github.com/Azure/azure-container-networking/log"
+	nns "github.com/Azure/azure-container-networking/proto/nodenetworkservice/3.302.0.744"
+	"google.golang.org/grpc"
+	"time"
+)
+
+const (
+	nnsPort = "6668"
```
Comment (ramiro-gamarra) [inline cni/nnsgrpcclient/nnsgrpclient.go:14]: how set in stone is this? should they ever change, code would need to as well. should this be config instead?

File: cni/nnsgrpcclient/nnsgrpclient.go [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+	apiTimeout = 5 * time.Minute // recommended timeout from node service
```
Comment (ramiro-gamarra) [inline cni/nnsgrpcclient/nnsgrpclient.go:15]: how intensive are these operations that they may require up to 5 minutes? 😁

**Review by ramiro-gamarra — COMMENTED** [review]
Some early feedback

File: cni/network/network.go [hunk 11]
```diff
@@ -791,6 +814,31 @@ func (plugin *netPlugin) Delete(args *cniSkel.CmdArgs) error {
...
+		err = plugin.nnsClient.DeleteContainerNetworking(context.Background(), k8sPodName, args.Netns)
+		return err
```
Comment (tamilmani1989) [inline cni/network/network.go:839]: do we return error for all cases even if container is already removed ? adding brief desc will be helpful

File: cni/nnsgrpcclient/nnsgrpclient.go [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+		ContainerId: podName,
+		NetworkNamespaceId: nwNamespace,
+	}
+
+	localCtx, cancel := context.WithTimeout(ctx, apiTimeout)
+	defer cancel()
+
+	res, err := client.ConfigureContainerNetworking(localCtx, req)
+	if err != nil {
+		log.Debugf("ConfigureContainerNetworking for container %s for nw namespace %s failed with error: %s",
```
Comment (tamilmani1989) [inline cni/nnsgrpcclient/nnsgrpclient.go:66]: one suggestion, for baremetal specific logs you can add prefix [baremetal]

File: cni/nnsgrpcclient/nnsgrpclient.go [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+    return err
+}
+
+func newNnsGrpcClient (ctx context.Context) (nns.NodeNetworkServiceClient, *grpc.ClientConn,  error)  {
+
+	localCtx, cancel := context.WithTimeout(ctx, connectionTimeout)
+	defer cancel()
+
+    nnsEndpoint := fmt.Sprintf(":%s", nnsPort)
+	conn, err := grpc.DialContext(
```
Comment (tamilmani1989) [inline cni/nnsgrpcclient/nnsgrpclient.go:82]: add log to print nnsendpoint cni connecting to

File: cni/nnsgrpcclient/nnsgrpclient.go [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+func configureContainerNetworking (
+	ctx context.Context,
+	reqtype nns.RequestType,
+	podName, nwNamespace string) error {
+
+	client, conn, err := newNnsGrpcClient(ctx)
+	if err != nil {
+		return err
```
Comment (tamilmani1989) [inline cni/nnsgrpcclient/nnsgrpclient.go:46]: hope we log this error somewhere

File: cni/nnsgrpcclient/nnsgrpclient.go [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+	"github.com/Azure/azure-container-networking/log"
+	nns "github.com/Azure/azure-container-networking/proto/nodenetworkservice/3.302.0.744"
+	"google.golang.org/grpc"
+	"time"
+)
+
+const (
+	nnsPort = "6668"
+	apiTimeout = 5 * time.Minute // recommended timeout from node service
+	connectionTimeout = 2 * time.Minute
```
Comment (tamilmani1989) [inline cni/nnsgrpcclient/nnsgrpclient.go:16]: is 2 mins isn't it longer timeout?

File: proto/nodenetworkservice/3.302.0.744/NodeNetworkService.proto [hunk 16]
```diff
@@ -0,0 +1,53 @@
+syntax = "proto3";
+package azure.nodenetworkservice;
```
Comment (tamilmani1989) [inline proto/nodenetworkservice/3.302.0.744/NodeNetworkService.proto:2]: can it be azurenodenetworkservice or azurenns

File: proto/nodenetworkservice/3.302.0.744/NodeNetworkService.proto [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+option go_package = ".;nodenetworkservice";
+
+service NodeNetworkService {
+    rpc ConfigureNetworking(ConfigureNetworkingRequest) returns (ConfigureNetworkingResponse);
```
Comment (tamilmani1989) [inline proto/nodenetworkservice/3.302.0.744/NodeNetworkService.proto:6]: can you add short description on what these api does?

also whats the diff between ConfigureNetworking and ConfigureContainerNetworking

File: cni/nnsgrpcclient/nnsgrpclient.go [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+			podName, nwNamespace, err)
+	} else {
+		log.Debugf("ConfigureContainerNetworking for container %s for nw namespace %s succeeded with result: %v",
+			podName, nwNamespace, res)
+	}
+
+    return err
+}
+
+func newNnsGrpcClient (ctx context.Context) (nns.NodeNetworkServiceClient, *grpc.ClientConn,  error)  {
```
Comment (tamilmani1989) [inline cni/nnsgrpcclient/nnsgrpclient.go:76]: short desc on what this function does would help

File: cni/nnsgrpcclient/nnsgrpclient.go [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+	req := &nns.ConfigureContainerNetworkingRequest{
+		RequestType: reqtype,
+		ContainerId: podName,
+		NetworkNamespaceId: nwNamespace,
+	}
+
+	localCtx, cancel := context.WithTimeout(ctx, apiTimeout)
+	defer cancel()
+
+	res, err := client.ConfigureContainerNetworking(localCtx, req)
```
Comment (tamilmani1989) [inline cni/nnsgrpcclient/nnsgrpclient.go:64]: can you add a comment on what this function does

**Review by tamilmani1989 — CHANGES_REQUESTED** [review]

File: cni/network/network.go [hunk 20]
```diff
@@ -60,13 +61,26 @@ const (
...
+)
+
 // NetPlugin represents the CNI network plugin.
 type netPlugin struct {
 	*cni.Plugin
 	nm          network.NetworkManager
 	ipamInvoker IPAMInvoker
 	report      *telemetry.CNIReport
 	tb          *telemetry.TelemetryBuffer
+	nnsClient   NnsClient
```
[the remark this answers is not in the document]
Comment (msvik) [reply cni/network/network.go:78]: Agree, this is an overhead. Having said that no one has brought up performance as an issue yet, so we took the path of least resistance for now. If performance become an issue due to this connection setup overhead, CNS can be considered an option.

File: cni/network/plugin/main.go (hunk 2 above)
Comment (msvik) [reply cni/network/plugin/main.go:177]: Moved it outside of plugin and renamed the package.

File: cni/nnsgrpcclient/nnsgrpcclient_test.go (hunk 4 above)
Comment (msvik) [reply cni/nnsgrpcclient/nnsgrpcclient_test.go:41]: ran gofmt on this file.

File: cni/nnsgrpcclient/nnsgrpcclient_test.go (hunk 3 above)
Comment (msvik) [reply cni/nnsgrpcclient/nnsgrpcclient_test.go:25]: fixed

File: cni/network/network.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (msvik) [reply cni/network/network.go:78]: I have moved the nns grpclient out of plugin. It is at the root level now just like other client packages. The reason it is part of netPlugin struct is that because we are doing a dependency injection here via an interface. And it is passed into plugin struct from main.go. This way we can also pass a mock implementation for testing. For example - we can add a parameter in CNI config saying - "UseMockNns" and when it is true, we initialize and pass a mock client all the way to plugin. That is the intention here.

File: cni/network/network.go [hunk 21]
```diff
@@ -60,13 +61,26 @@ const (
...
+}
+
+// client for node network service
+type NnsClient interface {
+	AddContainerNetworking(ctx context.Context, podName, nwNamespace string) error
+	DeleteContainerNetworking(ctx context.Context, podName, nwNamespace string) error
```
[the remark this answers is not in the document]
Comment (msvik) [reply cni/network/network.go]: Added documentation

File: cni/network/network.go (hunk 11 above)
Comment (msvik) [reply cni/network/network.go:846]: CNI is pass through. All cases are handled by node network service.

File: cni/nnsgrpcclient/nnsgrpcclient_test.go (hunk 6 above)
Comment (msvik) [reply cni/nnsgrpcclient/nnsgrpcclient_test.go:36]: Since this client is meant to be used on a host, it is written to reach out to a specific port on localhost. Test actually starts a mock nns server (which starts at the expected port), instantiates a client and invokes APIs

File: cni/nnsgrpcclient/nnsgrpcclient_test.go (hunk 5 above)
Comment (msvik) [reply cni/nnsgrpcclient/nnsgrpcclient_test.go:44]: make sense. resolved.

File: cni/nnsgrpcclient/nnsgrpcclient_test.go (hunk 8 above)
Comment (msvik) [reply cni/nnsgrpcclient/nnsgrpcclient_test.go:105]: client's connection timeout is 2 minutes, so it will retry for up 2 mins, so 25 seconds chosen here well within that range. If we run tests in parallel in this package, this could be a problem but in general we run it sequentially, so it shouldn't impact future tests.

File: cni/nnsgrpcclient/nnsgrpclient.go (hunk 9 above)
Comment (msvik) [reply cni/nnsgrpcclient/nnsgrpclient.go:14]: My understanding is that once it is set, these things don't really change since it is a host level component and will require deployment across the fleet for any client depending on it. Having it in the config is definitely better, but regardless we will need to produce a new package. I'll ask node service team if it can be changed in future.

[the remark this answers is not in the document]
Comment (msvik) [reply cni/nnsgrpcclient/nnsgrpclient.go:14]: done

File: cni/nnsgrpcclient/nnsgrpclient.go (hunk 10 above)
Comment (msvik) [reply cni/nnsgrpcclient/nnsgrpclient.go:15]: This was recommended by nns team to us. I don't think API is intensive or takes long time. When it works, whole thing gets done in matter of seconds. But so far we have only tested it on a single node with single pod being created. At a scale, it is possible they NNS might take longer. Scale testing hasn't be done yet.

File: cni/nnsgrpcclient/nnsgrpclient.go (hunk 14 above)
Comment (msvik) [reply cni/nnsgrpcclient/nnsgrpclient.go]: This is returned to plugin and plugin logs it just like any other error.

File: cni/nnsgrpcclient/nnsgrpclient.go (hunk 15 above)
Comment (msvik) [reply cni/nnsgrpcclient/nnsgrpclient.go:16]: default http connection timeout is 2 mins, so I kept it same.

File: cni/nnsgrpcclient/nnsgrpclient.go (hunk 19 above)
Comment (msvik) [reply cni/nnsgrpcclient/nnsgrpclient.go:64]: Added documentation in this package.

File: cni/nnsgrpcclient/nnsgrpclient.go (hunk 18 above)
Comment (msvik) [reply cni/nnsgrpcclient/nnsgrpclient.go:76]: done

File: cni/nnsgrpcclient/nnsgrpclient.go (hunk 12 above)
Comment (msvik) [reply cni/nnsgrpcclient/nnsgrpclient.go]: added the baremetal prefix in logs

File: proto/nodenetworkservice/3.302.0.744/NodeNetworkService.proto (hunk 16 above)
Comment (msvik) [reply proto/nodenetworkservice/3.302.0.744/NodeNetworkService.proto:2]: This comes from nns team, do I didn't want to rename it. When we generate go contracts from it, we override this package name to nodenetworkservice.

File: proto/nodenetworkservice/3.302.0.744/NodeNetworkService.proto (hunk 17 above)
Comment (msvik) [reply proto/nodenetworkservice/3.302.0.744/NodeNetworkService.proto:6]: Again, this file comes from nns team. in future it could come via a nuget too. So we don't want to edit it. I have already added description of ConfigureContainerNetwoking in grpc client. Other APIs for other services that consume this contract. We only consume ConfigureContainerNetworking API.

File: cni/nnsgrpcclient/nnsgrpcclient_test.go (hunk 7 above)
Comment (msvik) [reply cni/nnsgrpcclient/nnsgrpcclient_test.go:71]: This is not needed as such. I added it because same API in mock server returns result of both Add/Delete. So this extra check here makes sure of right operation was done on mock server.

File: cni/network/network.go (hunk 1 above)
Comment (msvik) [reply cni/network/network.go:844]: good suggestion. fixed

File: cni/network/network.go [hunk 22]
```diff
@@ -1143,3 +1192,42 @@ func determineSnat() (bool, bool, error) {
...
+					log.Printf("convertNnsToCniResult error: %s \n", err)
+					continue
+				}
+
+				gateway := net.ParseIP(ip.DefaultGateway)
+				ipConfig := &cniTypesCurr.IPConfig{
+					Address: *ipNet,
+					Gateway: gateway,
+					Version: ip.Version,
+					Interface: &intIndex,
```
Comment (ramiro-gamarra) [inline cni/network/network.go:1222]: do we own this struct? pointer to int is interesting

File: nns/grpcclient_test.go [hunk 23]
```diff
@@ -0,0 +1,107 @@
...
+	setup()
+	code := m.Run()
+	teardown()
+	os.Exit(code)
+}
+
+func setup() {
+	mockserver = nnsmockserver.NewNnsMockServer()
+	go mockserver.StartGrpcServer(nnsPort)
+	fmt.Println("mock nns server started\n")
```
Comment (ramiro-gamarra) [inline nns/grpcclient_test.go:25]: `fmt.Println` doesn't need line break. same for below

File: nns/grpcclient_test.go [hunk 24]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+// CNI ADD to add container to network - failure case
+func TestAddContainerNetworkingFailure(t *testing.T) {
+
+	client := &GrpcClient{}
+
+	var err error
+	if err, _ = client.AddContainerNetworking(context.Background(), "testpod", "testnwspace"); err == nil {
+		t.Fatalf("TestAddContainerNetworkingFailure failed. Expected error but none returned\n")
```
Comment (ramiro-gamarra) [inline nns/grpcclient_test.go:64]: line break not needed, same for below

**Review by ramiro-gamarra — COMMENTED** [review]
few more comments

File: cni/network/network.go [hunk 25]
```diff
@@ -60,13 +62,33 @@ const (
 	jsonFileExtension = ".json"
 )
 
+type  ExecutionModes string
```
[the remark this answers is not in the document]
Comment (msvik) [reply cni/network/network.go:65]: fixed

File: cni/network/network.go [hunk 26]
```diff
@@ -361,6 +379,24 @@ func (plugin *netPlugin) Add(args *cniSkel.CmdArgs) error {
 		return plugin.Errorf(errMsg)
 	}
 
+	log.Printf("Execution mode :%s", nwCfg.ExecutionMode)
+	if nwCfg.ExecutionMode == string(Baremetal) {
+		var  res *nnscontracts.ConfigureContainerNetworkingResponse
+        log.Printf("Baremetal mode. Calling vnet agent for ADD")
```
[the remark this answers is not in the document]
Comment (msvik) [reply cni/network/network.go:385]: I did gofmt on entire file

File: cni/network/network.go [hunk 27]
```diff
@@ -1143,3 +1192,42 @@ func determineSnat() (bool, bool, error) {
...
+	result := &cniTypesCurr.Result{}
+	var resultIpconfigs []*cniTypesCurr.IPConfig
+
+    if netRes.Interfaces != nil {
+		for i, ni :=  range netRes.Interfaces {
+
+			intIndex := i
+			for _, ip := range ni.Ipaddresses {
+
+				ipWithPrefix := strings.Join([]string { ip.Ip, ip.PrefixLength }, "/")
```
[the remark this answers is not in the document]
Comment (msvik) [reply cni/network/network.go:1210]: fixed

File: cni/network/network.go [hunk 28]
```diff
@@ -1143,3 +1192,42 @@ func determineSnat() (bool, bool, error) {
...
+				_, ipNet, err := net.ParseCIDR(ipWithPrefix)
+				if err != nil {
+					log.Printf("convertNnsToCniResult error: %s \n", err)
```
[the remark this answers is not in the document]
Comment (msvik) [reply cni/network/network.go:1213]: Added more info in the logs. Removed \n

File: cni/network/network.go (hunk 22 above)
Comment (msvik) [reply cni/network/network.go:1222]: This is vendored from container networking repo.

File: cni/network/network.go [hunk 29]
```diff
@@ -1143,3 +1193,47 @@ func determineSnat() (bool, bool, error) {
...
+	// This function does not add interfaces to CNI result. Reason being CRI (containerD in baremetal case)
+	// only looks for default interface named "eth0" and this default interface is added in the defer
+	// method of ADD method
+	result := &cniTypesCurr.Result{}
+	var resultIpconfigs []*cniTypesCurr.IPConfig
+
+	if netRes.Interfaces != nil {
+		for i, ni := range netRes.Interfaces {
+
+			intIndex := i
```
Comment (tamilmani1989) [inline cni/network/network.go:1212]: why do we need copy of this variable?

**Review by tamilmani1989 — CHANGES_REQUESTED** [review]

File: cni/network/network.go (hunk 29 above)
Comment (msvik) [reply cni/network/network.go:1212]: I think it being done when creating result object. That's why.

Comment (tamilmani1989) [reply cni/network/network.go:1212]: i think this is the place you were mentioning:` Interface: &intIndex,` . can't we use &i instead of &intIndex?

**Review by tamilmani1989 — APPROVED** [review]
